File: Android/FoodRescue/app/src/main/java/edu/ewu/team1/foodrescue/FoodEventAdapter.java
package edu.ewu.team1.foodrescue;

import android.content.Context;
import android.content.Intent;
import android.database.DataSetObserver;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListAdapter;
import android.widget.TextView;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import java.util.ArrayList;

public class FoodEventAdapter implements ListAdapter {
    private ArrayList<FoodEvent> events;
    private ArrayList<DataSetObserver> observers;
    private Context context;

    public FoodEventAdapter(ArrayList<FoodEvent> events, Context context) {
        this.events = events;
        observers = new ArrayList<>();
        this.context = context;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        if (convertView == null) {
            LayoutInflater inflater = (LayoutInflater) this.context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            assert inflater != null;
            convertView = inflater.inflate(R.layout.notification_item, parent);
        }
        FoodEvent event = events.get(position);
        ((TextView) convertView.findViewById(R.id.textViewTitle)).setText(event.title);
        ((TextView) convertView.findViewById(R.id.textViewBody)).setText(event.body);
        LatLng loc = new LatLng(event.getLat(), event.getLng());
        //Map
        MapView mapView = convertView.findViewById(R.id.mapViewEater);
        mapView.getMapAsync(googleMap -> {
            googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(loc, 17.0f));
            googleMap.addMarker(new MarkerOptions().position(loc).title(event.title));
        });

        convertView.findViewById(R.id.buttonClear).setOnClickListener(v -> {
            events.remove(position);
            for(DataSetObserver o : observers){
                o.notifyAll();
            }
        });

        convertView.findViewById(R.id.buttonDirections).setOnClickListener(v -> {
            Intent intent = new Intent(android.content.Intent.ACTION_VIEW,
                    Uri.parse("http://maps.google.com/maps?daddr=" + event.getLat() + "," + event.getLng()));
            context.startActivity(intent);
        });
        return convertView;
    }

    @Override
    public boolean areAllItemsEnabled() {
        return true;
    }

    @Override
    public boolean isEnabled(int position) {
        return true;
    }

    @Override
    public void registerDataSetObserver(DataSetObserver observer) {
        observers.add(observer);
    }

    @Override
    public void unregisterDataSetObserver(DataSetObserver observer) {
        observers.remove(observer);
    }

    @Override
    public int getCount() {
        return events.size();
    }

    @Override
    public Object getItem(int position) {
        return events.get(position);
    }

    @Override
    public long getItemId(int position) {
        return events.get(position).timestamp;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @Override
    public int getItemViewType(int position) {
        return IGNORE_ITEM_VIEW_TYPE;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    @Override
    public boolean isEmpty() {
        return false;
    }
}

File: Android/FoodRescue/app/src/main/java/edu/ewu/team1/foodrescue/FoodEventAdapter.kt
package edu.ewu.team1.foodrescue

import android.content.Context
import android.content.Intent
import android.database.DataSetObserver
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Adapter
import android.widget.ListAdapter
import android.widget.TextView
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import java.util.*

class FoodEventAdapter(private val events: ArrayList<FoodEvent>, private val context: Context) : ListAdapter {
	private val observers: ArrayList<DataSetObserver> = ArrayList()

	override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
		//Inflate the view if it isn't already
		val view =
				if (convertView == null) {
					val inflater = this.context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
					inflater.inflate(R.layout.notification_item, parent)
				} else {
					convertView
				}

		val event = events[position]
		(view!!.findViewById<View>(R.id.textViewTitle) as TextView).text = event.title
		(view.findViewById<View>(R.id.textViewBody) as TextView).text = event.body
		val loc = LatLng(event.lat, event.lng)
		//Map
		val mapView = view.findViewById<MapView>(R.id.mapViewEater)
		mapView.getMapAsync { googleMap ->
			googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(loc, 17.0f))
			googleMap.addMarker(MarkerOptions().position(loc).title(event.title))
		}

		view.findViewById<View>(R.id.buttonClear).setOnClickListener {
			events.removeAt(position)
			for (o in observers) {
				o.onChanged()
			}
		}

		view.findViewById<View>(R.id.buttonDirections).setOnClickListener {
			val intent = Intent(android.content.Intent.ACTION_VIEW,
					Uri.parse("http://maps.google.com/maps?daddr=" + event.lat + "," + event.lng))
			context.startActivity(intent)
		}
		return view
	}

	//region Useless garbage
	override fun areAllItemsEnabled(): Boolean {
		return true
	}

	override fun isEnabled(position: Int): Boolean {
		return true
	}

	override fun registerDataSetObserver(observer: DataSetObserver) {
		observers.add(observer)
	}

	override fun unregisterDataSetObserver(observer: DataSetObserver) {
		observers.remove(observer)
	}

	override fun getCount(): Int {
		return events.size
	}

	override fun getItem(position: Int): Any {
		return events[position]
	}

	override fun getItemId(position: Int): Long {
		return events[position].timestamp
	}

	override fun hasStableIds(): Boolean {
		return true
	}

	override fun getItemViewType(position: Int): Int {
		return Adapter.IGNORE_ITEM_VIEW_TYPE
	}

	override fun getViewTypeCount(): Int {
		return 1
	}

	override fun isEmpty(): Boolean {
		return false
	}
	//endregion
}
